### PR TITLE
ThrustFactor Updated

### PR DESCRIPTION
**Pull request by MajorCooke**
- Added ThrustFactor. Works just like DamageFactor, except individual named factors take precedence over the undefined. Default is 1.0.
- Added A_SetThrustFactor (float amount, damagetype, pointer). Leaving damagetype as "none" will set the global actor valuable.
- Added +NOIMPACTDAMAGE which prevents the actor from being flagged as blasted when thrusted.

### CODE REVIEW

File: src/p_mobj.cpp [hunk 1]
```diff
@@ -3163,6 +3167,53 @@ void AActor::SetRoll(angle_t r, bool interpolate)
 	}
 }
 
+void AActor::SetThrustFactor(FName dmgtype, fixed_t amount)
+{
+	if (!dmgtype || dmgtype == NAME_None || !stricmp("none", dmgtype) || !stricmp("normal", dmgtype))
```
Comment (Blzut3) [inline src/p_mobj.cpp]: You should be thinking twice when you stricmp an FName.  FNames are case insensitive so comparing to NAME_None is sufficient for the first one and much faster.  There's also NAME_Normal.

File: src/g_doom/a_archvile.cpp [hunk 2]
```diff
@@ -152,5 +154,6 @@ DEFINE_ACTION_FUNCTION_PARAMS(AActor, A_VileAttack)
 		P_RadiusAttack (fire, self, blastdmg, blastrad, dmgtype, 0);
 	}
 	if (!(target->flags7 & MF7_DONTTHRUST))
-		target->velz = Scale(thrust, 1000, target->Mass);
+		target->velz = Scale(thrust, 1000, target->Mass) * thrustmul;
```
Comment (Blzut3) [inline src/g_doom/a_archvile.cpp]: I keep seeing "* thrustmul" throughout.  Are you sure you're not missing a FixedMul on any of these?

Comment (MajorCooke) [inline src/g_doom/a_archvile.cpp]: 100% positive.

File: src/p_mobj.cpp (hunk 1 above)
Comment (MajorCooke) [inline src/p_mobj.cpp]: That's just it. I was ensuring case insensitivity was used.

But okay I can change those last two down to a simple 'dmgtype == NAME_Normal'

File: src/g_doom/a_archvile.cpp (hunk 2 above)
Comment (MajorCooke) [inline src/g_doom/a_archvile.cpp]: Yes. ThrustMul is merely a multiplier to how much the thrust being done is scaled by. Granted I could've been a bit more clever with the naming and just make it * thrustscale instead...

Comment (MajorCooke) [inline src/g_doom/a_archvile.cpp]: Yes. It's nothing more than a direct multiplier that amplifies or undermines a damagetype's damage.

Comment (Gaerzi) [inline src/g_doom/a_archvile.cpp]: "Thrustmul is already a fixed_t."

That would be why Blzut3 asked you that question, then. When you multiply a fixed_t by another fixed_t, you have to use a FixedMul. In the same way that if you multiply 50% by 50% you shouldn't get 2500% as the result...

**Comment by MajorCooke** [comment]
This one has been updated in a new pull request.

**Pull request closed without merging**